File: Cpp-CSharp-Blackmire/CSharpLibrary/Bar.cs
namespace CSharpFooBarLibrary
{
	public class Bar
	{
		private IBar impl;

		public void Register(IBar ibar)
		{
			this.impl = ibar;
		}

		public int PerformGetNumber()
		{
			if (impl == null)
			{
				return -1;
			}
			else
			{
				return impl.GetNumber();
			}
		}
	}
}

File: Cpp-CSharp-SharpCpp/CppConsoleProject/Bar.cpp
#include "Bar.hpp"

using namespace CSharpFooBarLibrary;

void Bar::Register(const std::shared_ptr<IBar>& ibar)
{
  this->impl = ibar;
}

int Bar::PerformGetNumber()
{
  if (this->impl == NULL) {
    return -1;
  } else {
    return impl->GetNumber();
  }
}

